src/glum/_glm.py

```diff
@@ -1328,6 +1334,295 @@ def predict(
+            Target values. Can be omitted if a covariance matrix has already
+            been computed.
+        mu : array-like, optional, default=None
+            Array with predictions. Estimated if absent.
+        offset : array-like, optional, default=None
+            Array with additive offsets.
+        sample_weight : array-like, shape (n_samples,), optional, default=None
+            Individual weights for each sample.
+        dispersion : float, optional, default=None
+            The dispersion parameter. Estimated if absent.
+        robust : boolean, optional, default=None
+            Whether to compute robust standard errors instead of normal ones.
+            If not specified, the model's ``robust`` attribute is used.
+        clusters : array-like, optional, default=None
+            Array with clusters membership. Clustered standard errors are
+            computed if clusters is not None.
+        expected_information : boolean, optional, default=None
+            Whether to use the expected or observed information matrix.
+            Only relevant when computing robust std-errors.
+            If not specified, the model's ``expected_information`` attribute is used.
+
+        Returns
+        -------
+        WaldTestResult
+            NamedTuple with test statistic, p-value, and degrees of freedom.
+        """
+
+        covariance_matrix = self.covariance_matrix(
+            X=X,
+            y=y,
+            mu=mu,
+            offset=offset,
+            sample_weight=sample_weight,
+            dispersion=dispersion,
+            robust=robust,
+            clusters=clusters,
+            expected_information=expected_information,
+        )
+
+        if self.fit_intercept:
+            beta = np.concatenate([[self.intercept_], self.coef_])
+        else:
+            beta = self.coef_
+
+        if r is None:
+            r = np.zeros(R.shape[0])
+
+        if R.shape[0] != r.shape[0]:
+            raise ValueError("R and r must have the same number of rows")
+        if R.shape[1] != beta.shape[0]:
+            raise ValueError("R must have one column for each coefficient")
+        # There is no point in checking that R is full rank. If it is not, then
+        # solve(RVR, Rb_r) will raise an exception.
+
+        beta = beta[:, np.newaxis]
+        r = r[:, np.newaxis]
+        Q = R.shape[0]
+
+        Rb_r = R @ beta - r  # R \beta - r
+        RVR = R @ covariance_matrix @ R.T  # R V R^T
+
+        # We want to calculate Rb_r^T (RVR)^{-1} Rb_r.
+        # We can do it in a more numerically stable way by using `scipy.linalg.solve`:
+        try:
+            test_stat = float(Rb_r.T @ linalg.solve(RVR, Rb_r))
+        except linalg.LinAlgError as err:
+            raise linalg.LinAlgError("The restriction matrix is not full rank") from err
+        p_value = 1 - stats.chi2.cdf(test_stat, Q)
+
+        return WaldTestResult(test_stat, p_value, Q)
+
+    def wald_test_feature_name(
```

Comment:
not a huge fan of the method name.
Here's a suggestion: combine the two methods into one called `wald_test`.
r and values have the same meaning. R and features could share the same argument. If the argument is a list of string, you know that the user wants to use feature name. If it's a numpy array of numeric type, you know the user want the "matrix" form.

To make the change, you could make the two methods private (start with an underscore), and have `wald_test` being a small function to check which one the user wants.

I'm also okay if you keep the current setup, but would recommend you think of a better method name for this one.

Reply:
I just saw that you are calling `wald_test_matrix` from this method. Maybe you could simply rename this method, and check at the start if the first argument is a numeric array, then directly call `_wald_test_matrix`.

Reply:
Good point, I'm not a big fan of the current names, either. Dispatching based on the type of `R` would work for the current two functions, but in `v3` I'm planning to add formula-based and term-name-based Wald-tests, and `R` would be a string / list of strings in those, too.

What do you think of an in-between solution? Have one public Wald-test function and make the current ones private, and the publc function would accept `R_matrix`, `features` (and later on `terms` and `formula`) parameters. Dispatch would be based on which of these is provided. We would avoid the weird names, but choosing how to test would remain a bit more explicit?

Reply:
Yeah, that's a very good solution!